src/snowflake/snowpark/_internal/data_source/datasource_partitioner.py

```diff
@@ -0,0 +1,229 @@
+#
+# Copyright (c) 2012-2025 Snowflake Computing Inc. All rights reserved.
+#
+
+import datetime
+import decimal
+from _decimal import ROUND_HALF_EVEN, ROUND_HALF_UP
+from typing import Optional, Union, List, Any, Type, Callable
+
+import pytz
+from dateutil import parser
+
+from snowflake.snowpark._internal.data_source.datasource_reader import DataSourceReader
+from snowflake.snowpark._internal.data_source.dbms_dialects.base_dialect import (
+    BaseDialect,
+)
+from snowflake.snowpark._internal.data_source.drivers.base_driver import BaseDriver
+
+import logging
+
+from snowflake.snowpark._internal.type_utils import type_string_to_type_object
+from snowflake.snowpark._internal.data_source.datasource_typing import Connection
+from snowflake.snowpark.types import (
+    StructType,
+    _NumericType,
+    DateType,
+    DataType,
+    TimestampType,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class DataSourcePartitioner:
+    def __init__(
+        self,
+        dialect_class: Type[BaseDialect],
+        driver_class: Type[BaseDriver],
+        create_connection: Callable[[], "Connection"],
+    ) -> None:
+        self.dialect = dialect_class()
+        self.driver = driver_class(create_connection)
+        self.driver_class = driver_class
+        self.dialect_class = dialect_class
+        self.create_connection = create_connection
+
+    def reader(self) -> DataSourceReader:
+        return DataSourceReader(self.driver_class, self.create_connection)
+
+    def schema(
+        self,
+        table_or_query: str,
+        custom_schema: Optional[Union[str, StructType]] = None,
+    ) -> StructType:
+        if custom_schema is None:
+            return self.driver.infer_schema_from_description(table_or_query)
+        else:
+            if isinstance(custom_schema, str):
+                schema = type_string_to_type_object(custom_schema)
+                if not isinstance(schema, StructType):
+                    logger.error(f"Invalid schema string: {custom_schema}")
+                    raise ValueError(
+                        f"Invalid schema string: {custom_schema}. "
+                        f"You should provide a valid schema string representing a struct type."
```

Comment:
I remember you has improved the error message in the main branch giving a example?